### PR TITLE
[14.0] [ADD] barcode_generator_product_variant

### PR DESCRIPTION
**Pull request by renda-dev**
Generate missing barcodes for a product's variants in bulk.

### CODE REVIEW

**Review by francesco-ooops — APPROVED** [review]
Functional ok!

@legalsylvain what do you think of this one?

File: barcode_generator_product_variant/models/barcode_rule.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+                default_rule = self.search(
+                    [("id", "!=", rec.id), ("is_default_rule", "=", True)]
+                )
+                if default_rule:
+                    raise UserError(
+                        _(
+                            "There is already another rule set as default: %(rule_name)s",
+                            rule_name=default_rule.name,
+                        )
+                    )
```
Comment (legalsylvain) [inline barcode_generator_product_variant/models/barcode_rule.py]: That is a huge part of code, to add a default that can be set natively on the wizard, via UI. Don't you think ?

File: barcode_generator_product_variant/wizards/product_product_barcode_wizard.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        )
+
+    product_tmpl_id = fields.Many2one(
+        "product.template", required=True, readonly=True, string="Product"
+    )
+    product_ids = fields.Many2many(
+        "product.product",
+        compute="_compute_product_ids",
+        store=True,
+        domain="[('product_tmpl_id', '=', product_tmpl_id)]",
```
Comment (legalsylvain) [inline barcode_generator_product_variant/wizards/product_product_barcode_wizard.py]:
```suggestion
```

store on transient are not necessary I guess. also domain on computed value is useless.

File: barcode_generator_product_variant/wizards/product_product_barcode_wizard.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    def action_show(self):
+        self.ensure_one()
+        return {
+            "type": "ir.actions.act_window",
+            "name": _("Barcode generation"),
+            "res_model": self._name,
+            "res_id": self.id,
+            "view_mode": "form",
+            "target": "new",
+        }
```
Comment (legalsylvain) [inline barcode_generator_product_variant/wizards/product_product_barcode_wizard.py]: action_show is unnecessary IMO. wizard can be call without that function.

```suggestion
`

File: barcode_generator_product_variant/wizards/product_product_barcode_wizard.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    product_ids = fields.Many2many(
+        "product.product",
+        compute="_compute_product_ids",
+        store=True,
+        domain="[('product_tmpl_id', '=', product_tmpl_id)]",
+        string="Variants",
+    )
+    barcode_rule_id = fields.Many2one(
+        string="Barcode Rule",
+        comodel_name="barcode.rule",
```
Comment (legalsylvain) [inline barcode_generator_product_variant/wizards/product_product_barcode_wizard.py:25]:
```suggestion
        comodel_name="barcode.rule",
        required=True,
```

File: barcode_generator_product_variant/wizards/product_product_barcode_wizard.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+            "res_model": self._name,
+            "res_id": self.id,
+            "view_mode": "form",
+            "target": "new",
+        }
+
+    def action_generate_barcodes(self):
+        self.ensure_one()
+        if not self.barcode_rule_id:
+            raise UserError(_("You must select a barcode rule to generate barcodes"))
```
Comment (legalsylvain) [inline barcode_generator_product_variant/wizards/product_product_barcode_wizard.py]: can be removed, if the field is required.

```suggestion

File: barcode_generator_product_variant/wizards/product_product_barcode_wizard.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+class ProductProductBarcodeWizard(models.TransientModel):
+    _name = "product.product.barcode.wizard"
+    _description = "Wizard to manage barcode generation "
+    "for variants of a product"
+
+    def _default_barcode_rule_id(self):
+        return self.env["barcode.rule"].search(
+            [("is_default_rule", "=", True)], limit=1
+        )
```
Comment (legalsylvain) [inline barcode_generator_product_variant/wizards/product_product_barcode_wizard.py]: (as said in barcode.rule file)

```suggestion

File: barcode_generator_product_variant/wizards/product_product_barcode_wizard.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        "product.product",
+        compute="_compute_product_ids",
+        store=True,
+        domain="[('product_tmpl_id', '=', product_tmpl_id)]",
+        string="Variants",
+    )
+    barcode_rule_id = fields.Many2one(
+        string="Barcode Rule",
+        comodel_name="barcode.rule",
+        default=_default_barcode_rule_id,
```
Comment (legalsylvain) [inline barcode_generator_product_variant/wizards/product_product_barcode_wizard.py:27]:
```suggestion
```

File: barcode_generator_product_variant/wizards/product_product_barcode_wizard.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        return [
+            (
+                rec.id,
+                _(
+                    "Barcode management for %(prod)s",
+                    prod=rec.product_tmpl_id.display_name,
+                ),
+            )
+            for rec in self
+        ]
```
Comment (legalsylvain) [inline barcode_generator_product_variant/wizards/product_product_barcode_wizard.py]: Naive question : never saw name_get for transientModel. Where it is used ?

File: barcode_generator_product_variant/wizards/product_product_barcode_wizard.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+    @api.depends("product_tmpl_id")
+    def _compute_product_ids(self):
+        for wiz in self:
+            if not wiz.product_tmpl_id:
+                wiz.product_ids = False
+                continue
+            wiz.product_ids = wiz.product_tmpl_id.product_variant_ids.filtered(
+                lambda x: not x.barcode
+            )
```
Comment (legalsylvain) [inline barcode_generator_product_variant/wizards/product_product_barcode_wizard.py]: nitpicking : simplification.

```suggestion
        for wizard in self:
            wizard.product_ids = wizard.mapped("product_tmpl_id.product_variant_ids").filtered(
                lambda x: not x.barcode
            )
```

File: barcode_generator_product_variant/readme/USAGE.rst [hunk 10]
```diff
@@ -0,0 +1,12 @@
+For barcode rule setup, please refer to readme of barcodes_generator_product. **Please note: only rules with "Base managed by Sequence" can be used for this flow.**
+
```
Comment (legalsylvain) [inline barcode_generator_product_variant/readme/USAGE.rst]: this module can be usefull also for base managed manually.
```suggestion
For barcode rule setup, please refer to readme of barcodes_generator_product.
```

File: barcode_generator_product_variant/wizards/product_product_barcode_wizard.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    _description = "Wizard to manage barcode generation "
+    "for variants of a product"
+
+    def _default_barcode_rule_id(self):
+        return self.env["barcode.rule"].search(
+            [("is_default_rule", "=", True)], limit=1
+        )
+
+    product_tmpl_id = fields.Many2one(
+        "product.template", required=True, readonly=True, string="Product"
```
Comment (legalsylvain) [inline barcode_generator_product_variant/wizards/product_product_barcode_wizard.py]:
```suggestion
        "product.template", required=True, string="Product"
```

**Review by legalsylvain — COMMENTED** [review]
Make sense ! thanks for sharing this module.

some remarks inline.

File: barcode_generator_product_variant/models/barcode_rule.py (hunk 1 above)
Comment (renda-dev) [reply barcode_generator_product_variant/models/barcode_rule.py]: I mean, it doesn't impact much on performance and it's a great feature in my opinion

File: barcode_generator_product_variant/wizards/product_product_barcode_wizard.py (hunk 8 above)
Comment (renda-dev) [reply barcode_generator_product_variant/wizards/product_product_barcode_wizard.py]: I think that is a leftover that i forgot, thanks!

**Comment by renda-dev** [comment]
Thank you for the huge review! @legalsylvain 

I've modified this PR according to the majority of the things that you pointed out, please let me know what you think :)

**Comment by francesco-ooops** [comment]
@legalsylvain ping :)

**Comment by legalsylvain** [comment]
Hi. Thanks for all the changes !

> I mean, it doesn't impact much on performance and it's a great feature in my opinion

Well, I just say that setting a default on a field is a native feature in odoo. So no need to add dozens of lines of code for that. Don't you think ? 

image

**Comment by francesco-ooops** [comment]
> Hi. Thanks for all the changes !
> 
> > I mean, it doesn't impact much on performance and it's a great feature in my opinion
> 
> Well, I just say that setting a default on a field is a native feature in odoo. So no need to add dozens of lines of code for that. Don't you think ?
> 
> image

Hi @legalsylvain , this feature you're referring to it's hardly user-friendly, requires developer mode access and is often not working as expected. A company with hundreds of employees cannot really rely on setting a default value for each user. 

On the contrary, it's quite common to add a default field in settings to specify a default value company-wide, but in this case there is no dependency on a specific app where we can add this "default" field, so we resorted to adding it on the record itself. 

I second @renda-dev opinion that it shouldn't be a big deal, so if it's not blocking we'd appreciate the approval :)

**Comment by legalsylvain** [comment]
> Hi @legalsylvain , this feature you're referring to it's hardly user-friendly, requires developer mode access and is often not working as expected. A company with hundreds of employees cannot really rely on setting a default value for each user.

Well, for me : 

> this feature you're referring to it's hardly user-friendly, requires developer mode access 

create and configure barcode rule is really not user-friendly. It has to be done by admin people.

> and is often not working as expected.

could you elaborate ?

> A company with hundreds of employees cannot really rely on setting a default value for each user.

The screenshot I did is not correct, I wanted to mention "all users" option. That way the admin instead of clicking on the new boolean, configure one time only, the ir.default. Not more work, just another way to configre.

That being said, it is not a blocking point ! You can go ahead if you think it's the good design.

regards.

File: barcode_generator_product_variant/tests/test_barcode_generator_product_variant.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        super().setUpClass()
+
+        cls.barcode_rule = cls.env.ref(
+            "barcode_generator_product_variant.product_generated_barcode_demo"
+        )
+        cls.test_product = cls.env.ref("product.product_product_4_product_template")
+
+    def test_generate_missing_barcodes(self):
+        self.assertFalse(
+            any(i for i in self.test_product.mapped("product_variant_ids.barcode"))
```
Comment (HekkiMelody) [inline barcode_generator_product_variant/tests/test_barcode_generator_product_variant.py]: **quibble (non-blocking)**: you don't need the generator here, `mapped` already returns a list

```suggestion
            any(self.test_product.mapped("product_variant_ids.barcode"))
```

File: barcode_generator_product_variant/tests/test_barcode_generator_product_variant.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+                "product_tmpl_id": self.test_product.id,
+            }
+        )
+
+        # Barcode rule should be the default one
+        self.assertEqual(self.barcode_rule, wizard.barcode_rule_id)
+
+        wizard.action_generate_barcodes()
+        self.assertTrue(
+            any(i for i in self.test_product.mapped("product_variant_ids.barcode"))
```
Comment (HekkiMelody) [inline barcode_generator_product_variant/tests/test_barcode_generator_product_variant.py]: **todo**: shouldn't this be `all`?

Also same remark as above, generator is unnecessary

```suggestion
            all(self.test_product.mapped("product_variant_ids.barcode"))
```

File: barcode_generator_product_variant/tests/test_barcode_generator_product_variant.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+        )
+        self.assertFalse(self.test_product.has_missing_barcodes)
+
+    def test_default_rule(self):
+        self.barcode_rule = self.env.ref(
+            "barcode_generator_product_variant.product_generated_barcode_demo"
+        )
```
Comment (HekkiMelody) [inline barcode_generator_product_variant/tests/test_barcode_generator_product_variant.py]: **todo**: `barcode_rule` is already assigned to self in `setUpClass`. Also, it's generally not a good idea to assign to `self` inside a test method, there's a risk of affecting other tests.

File: barcode_generator_product_variant/__manifest__.py [hunk 15]
```diff
@@ -0,0 +1,20 @@
+{
+    "name": "Barcode Generator Product Variant",
+    "version": "14.0.1.0.1",
```
Comment (HekkiMelody) [inline barcode_generator_product_variant/__manifest__.py]: **chore**: 

```suggestion
    "version": "14.0.1.0.0",
```

File: barcode_generator_product_variant/wizards/product_product_barcode_wizard_views.xml [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+        <field name="model">product.product.barcode.wizard</field>
+        <field name="arch" type="xml">
+            <form>
+                <sheet>
+                    <group>
+                        <field name="product_tmpl_id" />
+                    </group>
+                    <group name="Variants">
+                        <field name="product_ids" nolabel="1">
+                            <tree editable="bottom" default_order="default_code">
```
Comment (HekkiMelody) [inline barcode_generator_product_variant/wizards/product_product_barcode_wizard_views.xml]: **chore**: this field is `readonly` so the `editable` attribute does not have any effect and can be removed

```suggestion
                            <tree default_order="default_code">
```

**Review by HekkiMelody — CHANGES_REQUESTED** [review]
Generally ok, some minor code cleanup

**Comment by francesco-ooops** [comment]
> > and is often not working as expected.
> 
> could you elaborate ?

@legalsylvain I haven't used it lately, but seems to me not all fields can be set as default:

image

> The screenshot I did is not correct, I wanted to mention "all users" option. That way the admin instead of clicking on the new boolean, configure one time only, the ir.default. Not more work, just another way to configre.

I couldn't find a way to do that. Can you manage in runboat to set one barcode rule to be used in wizard using your suggestion? Thanks!

**Comment by HekkiMelody** [comment]
> > > and is often not working as expected.
> > 
> > 
> > could you elaborate ?
> 
> @legalsylvain I haven't used it lately, but seems to me not all fields can be set as default:
> 
> image
> 
> > The screenshot I did is not correct, I wanted to mention "all users" option. That way the admin instead of clicking on the new boolean, configure one time only, the ir.default. Not more work, just another way to configre.
> 
> I couldn't find a way to do that. Can you manage in runboat to set one barcode rule to be used in wizard using your suggestion? Thanks!

You are trying to set the default on the product model, instead of the wizard model.

1. Click "Manage missing barcodes"
2. Click the debug symbol on the top left of the popup 

image

3. Click "Set Default" there.

**Comment by francesco-ooops** [comment]
Ok, this is good to know for someone who implements, but terrible UX if you want a responsible user (not "any" user) to be independent in setting the default barcode rule IMHO. I would maintain the current implementation.

**Review by HekkiMelody — APPROVED** [review]
Code review

**Comment by francesco-ooops** [comment]
@legalsylvain could I ask for approval? :)

File: barcode_generator_product_variant/models/barcode_rule.py (hunk 1 above)
Comment (rousseldenis) [reply barcode_generator_product_variant/models/barcode_rule.py]: Apart of that, I find this could be defined in barcode_generator_product.

**Comment by legalsylvain** [comment]
Hi @francesco-ooops. the PR is still approved, and I have no merge right on that repo. 
(https://github.com/OCA/repo-maintainer-conf/blob/master/conf/psc/logistics.yml)

thanks.

**Comment by francesco-ooops** [comment]
sure, I didn't ask for merging, just for the tick as you reviewed the module, but no biggie :)

image

**Comment by renda-dev** [comment]
Hi, I've moved the "Default Rule" option in the `barcode_generator_abstract` module, and I personally think that it's a great feature (299b11cdd54120dcb56b14f970e7eff5bb76f7a9)

Please, let me know what do you think about that :)
@legalsylvain @rousseldenis

File: barcode_generator_product_variant/views/barcode_rule_views.xml [hunk 17]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<odoo>
+    <record model="ir.ui.view" id="view_barcode_rule_form">
```
Comment (legalsylvain) [inline barcode_generator_product_variant/views/barcode_rule_views.xml]: Please move also this view in the abstract module.

File: barcodes_generator_abstract/models/barcode_rule.py [hunk 18]
```diff
@@ -49,6 +50,14 @@ class BarcodeRule(models.Model):
         "a new record in the mixed model.",
     )
 
+    is_default_rule = fields.Boolean("Is Default Barcode Rule", copy=False)
```
Comment (legalsylvain) [inline barcodes_generator_abstract/models/barcode_rule.py]:
```suggestion
    is_default = fields.Boolean("Is Default Barcode Rule", copy=False)
```
suggestion : at rule level, I think 'is default is enough'.

could you also add an "help" param to mention what this checkbox does ?

File: barcodes_generator_abstract/models/barcode_rule.py [hunk 19]
```diff
@@ -49,6 +50,14 @@ class BarcodeRule(models.Model):
...
+
+    @api.constrains("is_default_rule")
+    def _check_is_default_rule(self):
+        default_rule = self.search_count([("is_default_rule", "=", True)])
+        if default_rule > 1:
+            raise UserError(_("Only one rule can be set as default."))
```
Comment (legalsylvain) [inline barcodes_generator_abstract/models/barcode_rule.py]: This check looks wrong. 
We can have a default rule for partner model, and another one for product model.

**Review by legalsylvain — CHANGES_REQUESTED** [review]
Some remarks inline. Otherwise, LGTM. Thanks !

**Review by legalsylvain — APPROVED** [review]
Thanks !

**Comment by francesco-ooops** [comment]
@rousseldenis good to go? :)

**Pull request merged**